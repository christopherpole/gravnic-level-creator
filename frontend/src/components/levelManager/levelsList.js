import React, { Component } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { SortableContainer } from 'react-sortable-hoc';

import Level from './level';
import { getSortedLevels } from '../../selectors';
import { retrieveLevels, reorderLevels } from '../../actions/levelManager';
import LoadingIcon from '../common/loadingIcon';
import WarningIcon from '../common/warningIcon';

export const Wrapper = styled.div`
  flex-grow: 1;
  position: relative;
`;

export const ScrollableArea = styled.div`
  top: 0;
  right: 0;
  left: 0;
  bottom: 0;
  position: absolute;
`;

export const ListWrapper = SortableContainer(styled.ul`
  padding: ${props => props.theme.structureSpacing};
  overflow-y: scroll;
  max-height: 100%;
`);

export const FullContainer = styled.div`
  display: flex;
  align-items: center;
  flex-direction: column;
  justify-content: center;
  text-align: center;
  position: absolute;
  top: ${props => props.theme.structureSpacing};
  left: ${props => props.theme.structureSpacing};
  bottom: ${props => props.theme.structureSpacing};
  right: ${props => props.theme.structureSpacing};
`;

export const LoadingContainer = styled(FullContainer)``;

export const ErrorContainer = styled(FullContainer)``;

export const NoLevelsContainer = styled(FullContainer)``;

export const NoLevelsCopy = styled.p``;

export const ErrorMessage = styled.p`
  margin-top: calc(${props => props.theme.structureSpacing} / 2);
`;

export class LevelsList extends Component {
  componentDidMount() {
    this.props.retrieveLevelsAction();
  }

  render() {
    const {
      loading,
      error,
      sortedLevels,
      selectedLevelId,
      currentLevelId,
      renamingLevelId,
      renamingLevelName,
      reorderLevelsAction,
    } = this.props;

    if (loading) {
      return (
        <Wrapper>
          <LoadingContainer>
            <LoadingIcon />
          </LoadingContainer>
        </Wrapper>
      );
    }

    if (error) {
      return (
        <Wrapper>
          <ErrorContainer>
            <WarningIcon />
            <ErrorMessage>There was a problem communicating with the server</ErrorMessage>
          </ErrorContainer>
        </Wrapper>
      );
    }

    if (!sortedLevels.length) {
      return (
        <Wrapper>
          <NoLevelsContainer>
            <NoLevelsCopy className="no-levels">No levels found</NoLevelsCopy>
          </NoLevelsContainer>
        </Wrapper>
      );
    }

    return (
      <Wrapper>
        <ScrollableArea>
          <ListWrapper
            lockAxis="y"
            helperClass="drag-cursor"
            lockToContainerEdges
            distance={5}
            onSortEnd={({ oldIndex, newIndex }) => {
              reorderLevelsAction(oldIndex, newIndex);
            }}
          >
            {sortedLevels.map((level, index) => (
              <Level
                {...level}
                key={level.id}
                index={index}
                isSelected={selectedLevelId && level.id === selectedLevelId}
                isCurrent={currentLevelId && level.id === currentLevelId}
                renamingValue={
                  renamingLevelId && level.id === renamingLevelId ? renamingLevelName : null
                }
              >
                {level.name}
              </Level>
            ))}
          </ListWrapper>
        </ScrollableArea>
      </Wrapper>
    );
  }
}

LevelsList.defaultProps = {
  selectedLevelId: null,
  currentLevelId: null,
  renamingLevelId: null,
  renamingLevelName: null,
};

LevelsList.propTypes = {
  loading: PropTypes.bool.isRequired,
  error: PropTypes.bool.isRequired,
  sortedLevels: PropTypes.array.isRequired,
  selectedLevelId: PropTypes.string,
  currentLevelId: PropTypes.string,
  renamingLevelId: PropTypes.string,
  renamingLevelName: PropTypes.string,
  retrieveLevelsAction: PropTypes.func.isRequired,
  reorderLevelsAction: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  loading: state.levelManager.loading,
  error: state.levelManager.error,
  selectedLevelId: state.levelManager.selectedLevelId,
  currentLevelId: state.levelManager.currentLevelId,
  renamingLevelId: state.levelManager.renamingLevelId,
  renamingLevelName: state.levelManager.renamingLevelName,
  sortedLevels: getSortedLevels(state),
});

const mapDispatchToProps = dispatch => ({
  retrieveLevelsAction: bindActionCreators(retrieveLevels, dispatch),
  reorderLevelsAction: bindActionCreators(reorderLevels, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(LevelsList);
