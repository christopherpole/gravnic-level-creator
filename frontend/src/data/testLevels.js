import { MOVE_DOWN, MOVE_UP } from 'gravnic-game';

export default [
  {
    id: '1',
    name: 'Test level 1',
    solution: [MOVE_DOWN, MOVE_UP],
    tiles: [
      {
        position: 0,
        selectedTileId: '1',
      },
      {
        position: 1,
        selectedTileId: '1',
      },
      {
        position: 2,
        selectedTileId: '1',
      },
      {
        position: 3,
        selectedTileId: '1',
      },
      {
        position: 4,
        selectedTileId: '1',
      },
      {
        position: 5,
        selectedTileId: '1',
      },
      {
        position: 6,
        selectedTileId: '1',
      },
      {
        position: 7,
        selectedTileId: '1',
      },
      {
        position: 8,
        selectedTileId: '1',
      },
      {
        position: 9,
        selectedTileId: '1',
      },
      {
        position: 10,
        selectedTileId: '1',
      },
      {
        position: 11,
        selectedTileId: '1',
      },
      {
        position: 12,
        selectedTileId: '1',
      },
      {
        position: 13,
        selectedTileId: '1',
      },
      {
        position: 14,
        selectedTileId: '1',
      },
      {
        position: 15,
        selectedTileId: '1',
      },
      {
        position: 16,
        selectedTileId: '1',
      },
      {
        position: 17,
        selectedTileId: '1',
      },
      {
        position: 18,
        selectedTileId: '1',
      },
      {
        position: 19,
        selectedTileId: '1',
      },
      {
        position: 20,
        selectedTileId: '1',
      },
      {
        position: 21,
        selectedTileId: '1',
      },
      {
        position: 22,
        selectedTileId: '1',
      },
      {
        position: 23,
        selectedTileId: '2',
      },
      {
        position: 24,
        selectedTileId: '1',
      },
      {
        position: 25,
        selectedTileId: '1',
      },
      {
        position: 26,
        selectedTileId: '1',
      },
      {
        position: 27,
        selectedTileId: '1',
      },
      {
        position: 28,
        selectedTileId: '1',
      },
      {
        position: 29,
        selectedTileId: '1',
      },
      {
        position: 30,
        selectedTileId: '1',
      },
      {
        position: 31,
        selectedTileId: '1',
      },
      {
        position: 32,
        selectedTileId: '2',
      },
      {
        position: 33,
        selectedTileId: '2',
      },
      {
        position: 34,
        selectedTileId: '2',
      },
      {
        position: 35,
        selectedTileId: '2',
      },
      {
        position: 36,
        selectedTileId: '2',
      },
      {
        position: 37,
        selectedTileId: '2',
      },
      {
        position: 38,
        selectedTileId: '1',
      },
      {
        position: 39,
        selectedTileId: '1',
      },
      {
        position: 40,
        selectedTileId: '1',
      },
      {
        position: 41,
        selectedTileId: '1',
      },
      {
        position: 42,
        selectedTileId: '2',
      },
      {
        position: 43,
        selectedTileId: '2',
      },
      {
        position: 44,
        selectedTileId: '2',
      },
      {
        position: 45,
        selectedTileId: '2',
      },
      {
        position: 46,
        selectedTileId: '2',
      },
      {
        position: 47,
        selectedTileId: '2',
      },
      {
        position: 48,
        selectedTileId: '4',
      },
      {
        position: 49,
        selectedTileId: '1',
      },
      {
        position: 50,
        selectedTileId: '1',
      },
      {
        position: 51,
        selectedTileId: '1',
      },
      {
        position: 52,
        selectedTileId: '2',
      },
      {
        position: 53,
        selectedTileId: '2',
      },
      {
        position: 54,
        selectedTileId: '2',
      },
      {
        position: 55,
        selectedTileId: '2',
      },
      {
        position: 56,
        selectedTileId: '2',
      },
      {
        position: 57,
        selectedTileId: '1',
      },
      {
        position: 58,
        selectedTileId: '1',
      },
      {
        position: 59,
        selectedTileId: '1',
      },
      {
        position: 60,
        selectedTileId: '1',
      },
      {
        position: 61,
        selectedTileId: '1',
      },
      {
        position: 62,
        selectedTileId: '2',
      },
      {
        position: 63,
        selectedTileId: '2',
      },
      {
        position: 64,
        selectedTileId: '2',
      },
      {
        position: 65,
        selectedTileId: '2',
      },
      {
        position: 66,
        selectedTileId: '1',
      },
      {
        position: 67,
        selectedTileId: '1',
      },
      {
        position: 68,
        selectedTileId: '1',
      },
      {
        position: 69,
        selectedTileId: '1',
      },
      {
        position: 70,
        selectedTileId: '1',
      },
      {
        position: 71,
        selectedTileId: '1',
      },
      {
        position: 72,
        selectedTileId: '1',
      },
      {
        position: 73,
        selectedTileId: '1',
      },
      {
        position: 74,
        selectedTileId: '1',
      },
      {
        position: 75,
        selectedTileId: '1',
      },
      {
        position: 76,
        selectedTileId: '1',
      },
      {
        position: 77,
        selectedTileId: '1',
      },
      {
        position: 78,
        selectedTileId: '1',
      },
      {
        position: 79,
        selectedTileId: '1',
      },
      {
        position: 80,
        selectedTileId: '1',
      },
      {
        position: 81,
        selectedTileId: '1',
      },
      {
        position: 82,
        selectedTileId: '1',
      },
      {
        position: 83,
        selectedTileId: '1',
      },
      {
        position: 84,
        selectedTileId: '1',
      },
      {
        position: 85,
        selectedTileId: '1',
      },
      {
        position: 86,
        selectedTileId: '1',
      },
      {
        position: 87,
        selectedTileId: '1',
      },
      {
        position: 88,
        selectedTileId: '1',
      },
      {
        position: 89,
        selectedTileId: '1',
      },
      {
        position: 90,
        selectedTileId: '1',
      },
      {
        position: 91,
        selectedTileId: '1',
      },
      {
        position: 92,
        selectedTileId: '1',
      },
      {
        position: 93,
        selectedTileId: '1',
      },
      {
        position: 94,
        selectedTileId: '1',
      },
      {
        position: 95,
        selectedTileId: '1',
      },
      {
        position: 96,
        selectedTileId: '1',
      },
      {
        position: 97,
        selectedTileId: '1',
      },
      {
        position: 98,
        selectedTileId: '1',
      },
      {
        position: 99,
        selectedTileId: '1',
      },
    ],
    stars: [1, 2, 3],
    position: 1,
  },
  {
    id: '2',
    name: 'Test level 2',
    solution: [MOVE_UP],
    tiles: [
      {
        position: 0,
        selectedTileId: '1',
      },
      {
        position: 1,
        selectedTileId: '1',
      },
      {
        position: 2,
        selectedTileId: '1',
      },
      {
        position: 3,
        selectedTileId: '1',
      },
      {
        position: 4,
        selectedTileId: '1',
      },
      {
        position: 5,
        selectedTileId: '1',
      },
      {
        position: 6,
        selectedTileId: '1',
      },
      {
        position: 7,
        selectedTileId: '1',
      },
      {
        position: 8,
        selectedTileId: '1',
      },
      {
        position: 9,
        selectedTileId: '1',
      },
      {
        position: 10,
        selectedTileId: '1',
      },
      {
        position: 11,
        selectedTileId: '1',
      },
      {
        position: 12,
        selectedTileId: '1',
      },
      {
        position: 13,
        selectedTileId: '1',
      },
      {
        position: 14,
        selectedTileId: '1',
      },
      {
        position: 15,
        selectedTileId: '1',
      },
      {
        position: 16,
        selectedTileId: '1',
      },
      {
        position: 17,
        selectedTileId: '1',
      },
      {
        position: 18,
        selectedTileId: '1',
      },
      {
        position: 19,
        selectedTileId: '1',
      },
      {
        position: 20,
        selectedTileId: '1',
      },
      {
        position: 21,
        selectedTileId: '1',
      },
      {
        position: 22,
        selectedTileId: '1',
      },
      {
        position: 23,
        selectedTileId: '2',
      },
      {
        position: 24,
        selectedTileId: '1',
      },
      {
        position: 25,
        selectedTileId: '1',
      },
      {
        position: 26,
        selectedTileId: '1',
      },
      {
        position: 27,
        selectedTileId: '1',
      },
      {
        position: 28,
        selectedTileId: '1',
      },
      {
        position: 29,
        selectedTileId: '1',
      },
      {
        position: 30,
        selectedTileId: '1',
      },
      {
        position: 31,
        selectedTileId: '1',
      },
      {
        position: 32,
        selectedTileId: '2',
      },
      {
        position: 33,
        selectedTileId: '2',
      },
      {
        position: 34,
        selectedTileId: '2',
      },
      {
        position: 35,
        selectedTileId: '2',
      },
      {
        position: 36,
        selectedTileId: '2',
      },
      {
        position: 37,
        selectedTileId: '2',
      },
      {
        position: 38,
        selectedTileId: '1',
      },
      {
        position: 39,
        selectedTileId: '1',
      },
      {
        position: 40,
        selectedTileId: '1',
      },
      {
        position: 41,
        selectedTileId: '1',
      },
      {
        position: 42,
        selectedTileId: '2',
      },
      {
        position: 43,
        selectedTileId: '2',
      },
      {
        position: 44,
        selectedTileId: '2',
      },
      {
        position: 45,
        selectedTileId: '2',
      },
      {
        position: 46,
        selectedTileId: '2',
      },
      {
        position: 47,
        selectedTileId: '4',
      },
      {
        position: 48,
        selectedTileId: '2',
      },
      {
        position: 49,
        selectedTileId: '1',
      },
      {
        position: 50,
        selectedTileId: '2',
      },
      {
        position: 51,
        selectedTileId: '2',
      },
      {
        position: 52,
        selectedTileId: '2',
      },
      {
        position: 53,
        selectedTileId: '2',
      },
      {
        position: 54,
        selectedTileId: '2',
      },
      {
        position: 55,
        selectedTileId: '2',
      },
      {
        position: 56,
        selectedTileId: '2',
      },
      {
        position: 57,
        selectedTileId: '1',
      },
      {
        position: 58,
        selectedTileId: '1',
      },
      {
        position: 59,
        selectedTileId: '1',
      },
      {
        position: 60,
        selectedTileId: '1',
      },
      {
        position: 61,
        selectedTileId: '1',
      },
      {
        position: 62,
        selectedTileId: '2',
      },
      {
        position: 63,
        selectedTileId: '2',
      },
      {
        position: 64,
        selectedTileId: '2',
      },
      {
        position: 65,
        selectedTileId: '2',
      },
      {
        position: 66,
        selectedTileId: '1',
      },
      {
        position: 67,
        selectedTileId: '1',
      },
      {
        position: 68,
        selectedTileId: '1',
      },
      {
        position: 69,
        selectedTileId: '1',
      },
      {
        position: 70,
        selectedTileId: '1',
      },
      {
        position: 71,
        selectedTileId: '1',
      },
      {
        position: 72,
        selectedTileId: '1',
      },
      {
        position: 73,
        selectedTileId: '1',
      },
      {
        position: 74,
        selectedTileId: '1',
      },
      {
        position: 75,
        selectedTileId: '1',
      },
      {
        position: 76,
        selectedTileId: '1',
      },
      {
        position: 77,
        selectedTileId: '1',
      },
      {
        position: 78,
        selectedTileId: '1',
      },
      {
        position: 79,
        selectedTileId: '1',
      },
      {
        position: 80,
        selectedTileId: '1',
      },
      {
        position: 81,
        selectedTileId: '1',
      },
      {
        position: 82,
        selectedTileId: '1',
      },
      {
        position: 83,
        selectedTileId: '1',
      },
      {
        position: 84,
        selectedTileId: '1',
      },
      {
        position: 85,
        selectedTileId: '1',
      },
      {
        position: 86,
        selectedTileId: '1',
      },
      {
        position: 87,
        selectedTileId: '1',
      },
      {
        position: 88,
        selectedTileId: '1',
      },
      {
        position: 89,
        selectedTileId: '1',
      },
      {
        position: 90,
        selectedTileId: '1',
      },
      {
        position: 91,
        selectedTileId: '1',
      },
      {
        position: 92,
        selectedTileId: '1',
      },
      {
        position: 93,
        selectedTileId: '1',
      },
      {
        position: 94,
        selectedTileId: '1',
      },
      {
        position: 95,
        selectedTileId: '1',
      },
      {
        position: 96,
        selectedTileId: '1',
      },
      {
        position: 97,
        selectedTileId: '1',
      },
      {
        position: 98,
        selectedTileId: '1',
      },
      {
        position: 99,
        selectedTileId: '1',
      },
    ],
    stars: [1, 2, 3],
    position: 2,
  },
  {
    id: '3',
    name: 'Test level 3',
    solution: [],
    tiles: [
      {
        position: 0,
        selectedTileId: '1',
      },
      {
        position: 1,
        selectedTileId: '1',
      },
      {
        position: 2,
        selectedTileId: '1',
      },
      {
        position: 3,
        selectedTileId: '1',
      },
      {
        position: 4,
        selectedTileId: '1',
      },
      {
        position: 5,
        selectedTileId: '1',
      },
      {
        position: 6,
        selectedTileId: '1',
      },
      {
        position: 7,
        selectedTileId: '1',
      },
      {
        position: 8,
        selectedTileId: '1',
      },
      {
        position: 9,
        selectedTileId: '1',
      },
      {
        position: 10,
        selectedTileId: '1',
      },
      {
        position: 11,
        selectedTileId: '1',
      },
      {
        position: 12,
        selectedTileId: '1',
      },
      {
        position: 13,
        selectedTileId: '1',
      },
      {
        position: 14,
        selectedTileId: '1',
      },
      {
        position: 15,
        selectedTileId: '1',
      },
      {
        position: 16,
        selectedTileId: '1',
      },
      {
        position: 17,
        selectedTileId: '1',
      },
      {
        position: 18,
        selectedTileId: '1',
      },
      {
        position: 19,
        selectedTileId: '1',
      },
      {
        position: 20,
        selectedTileId: '1',
      },
      {
        position: 21,
        selectedTileId: '1',
      },
      {
        position: 22,
        selectedTileId: '1',
      },
      {
        position: 23,
        selectedTileId: '2',
      },
      {
        position: 24,
        selectedTileId: '1',
      },
      {
        position: 25,
        selectedTileId: '1',
      },
      {
        position: 26,
        selectedTileId: '1',
      },
      {
        position: 27,
        selectedTileId: '1',
      },
      {
        position: 28,
        selectedTileId: '1',
      },
      {
        position: 29,
        selectedTileId: '1',
      },
      {
        position: 30,
        selectedTileId: '1',
      },
      {
        position: 31,
        selectedTileId: '1',
      },
      {
        position: 32,
        selectedTileId: '2',
      },
      {
        position: 33,
        selectedTileId: '2',
      },
      {
        position: 34,
        selectedTileId: '2',
      },
      {
        position: 35,
        selectedTileId: '2',
      },
      {
        position: 36,
        selectedTileId: '2',
      },
      {
        position: 37,
        selectedTileId: '2',
      },
      {
        position: 38,
        selectedTileId: '1',
      },
      {
        position: 39,
        selectedTileId: '1',
      },
      {
        position: 40,
        selectedTileId: '1',
      },
      {
        position: 41,
        selectedTileId: '1',
      },
      {
        position: 42,
        selectedTileId: '2',
      },
      {
        position: 43,
        selectedTileId: '2',
      },
      {
        position: 44,
        selectedTileId: '2',
      },
      {
        position: 45,
        selectedTileId: '4',
      },
      {
        position: 46,
        selectedTileId: '2',
      },
      {
        position: 47,
        selectedTileId: '2',
      },
      {
        position: 48,
        selectedTileId: '2',
      },
      {
        position: 49,
        selectedTileId: '2',
      },
      {
        position: 50,
        selectedTileId: '2',
      },
      {
        position: 51,
        selectedTileId: '1',
      },
      {
        position: 52,
        selectedTileId: '1',
      },
      {
        position: 53,
        selectedTileId: '2',
      },
      {
        position: 54,
        selectedTileId: '2',
      },
      {
        position: 55,
        selectedTileId: '2',
      },
      {
        position: 56,
        selectedTileId: '2',
      },
      {
        position: 57,
        selectedTileId: '1',
      },
      {
        position: 58,
        selectedTileId: '1',
      },
      {
        position: 59,
        selectedTileId: '1',
      },
      {
        position: 60,
        selectedTileId: '1',
      },
      {
        position: 61,
        selectedTileId: '1',
      },
      {
        position: 62,
        selectedTileId: '2',
      },
      {
        position: 63,
        selectedTileId: '2',
      },
      {
        position: 64,
        selectedTileId: '2',
      },
      {
        position: 65,
        selectedTileId: '2',
      },
      {
        position: 66,
        selectedTileId: '1',
      },
      {
        position: 67,
        selectedTileId: '1',
      },
      {
        position: 68,
        selectedTileId: '1',
      },
      {
        position: 69,
        selectedTileId: '1',
      },
      {
        position: 70,
        selectedTileId: '1',
      },
      {
        position: 71,
        selectedTileId: '1',
      },
      {
        position: 72,
        selectedTileId: '1',
      },
      {
        position: 73,
        selectedTileId: '1',
      },
      {
        position: 74,
        selectedTileId: '1',
      },
      {
        position: 75,
        selectedTileId: '1',
      },
      {
        position: 76,
        selectedTileId: '1',
      },
      {
        position: 77,
        selectedTileId: '1',
      },
      {
        position: 78,
        selectedTileId: '1',
      },
      {
        position: 79,
        selectedTileId: '1',
      },
      {
        position: 80,
        selectedTileId: '1',
      },
      {
        position: 81,
        selectedTileId: '1',
      },
      {
        position: 82,
        selectedTileId: '1',
      },
      {
        position: 83,
        selectedTileId: '1',
      },
      {
        position: 84,
        selectedTileId: '1',
      },
      {
        position: 85,
        selectedTileId: '1',
      },
      {
        position: 86,
        selectedTileId: '1',
      },
      {
        position: 87,
        selectedTileId: '1',
      },
      {
        position: 88,
        selectedTileId: '1',
      },
      {
        position: 89,
        selectedTileId: '1',
      },
      {
        position: 90,
        selectedTileId: '1',
      },
      {
        position: 91,
        selectedTileId: '1',
      },
      {
        position: 92,
        selectedTileId: '1',
      },
      {
        position: 93,
        selectedTileId: '1',
      },
      {
        position: 94,
        selectedTileId: '1',
      },
      {
        position: 95,
        selectedTileId: '1',
      },
      {
        position: 96,
        selectedTileId: '1',
      },
      {
        position: 97,
        selectedTileId: '1',
      },
      {
        position: 98,
        selectedTileId: '1',
      },
      {
        position: 99,
        selectedTileId: '1',
      },
    ],
    stars: [1, 2, 3],
    position: 3,
  },
];
