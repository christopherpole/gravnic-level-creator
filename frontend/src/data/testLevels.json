[
  {
    "id": "1",
    "name": "Test level 1",
    "tiles": [
      {
        "position": 0,
        "selectedTileId": 2
      },
      {
        "position": 1,
        "selectedTileId": 3
      },
      {
        "position": 2,
        "selectedTileId": 0
      }
    ],
    "stars": [1, 2, 3]
  },
  {
    "id": "2",
    "name": "Test level 2",
    "tiles": [
      {
        "position": 0,
        "selectedTileId": 2
      },
      {
        "position": 1,
        "selectedTileId": 3
      },
      {
        "position": 2,
        "selectedTileId": 0
      }
    ],
    "stars": [1, 2, 3]
  },
  {
    "id": "3",
    "name": "Test level 3",
    "tiles": [
      {
        "position": 0,
        "selectedTileId": 2
      },
      {
        "position": 1,
        "selectedTileId": 3
      },
      {
        "position": 2,
        "selectedTileId": 0
      }
    ],
    "stars": [1, 2, 3]
  }
]
