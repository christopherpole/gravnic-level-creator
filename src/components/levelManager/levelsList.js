import React from 'react';
import styled from 'styled-components';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import Level from './level';

export const Wrapper = styled.ul`
  padding: ${props => props.theme.structureSpacing};
  flex-grow: 1;
`;

const LevelsList = ({ levels, selectedLevelId, currentLevelId }) => (
  <Wrapper>
    {levels.map(level => (
      <Level
        {...level}
        key={level.id}
        isSelected={selectedLevelId && level.id === selectedLevelId}
        isCurrent={currentLevelId && level.id === currentLevelId}
      >
        {level.name}
      </Level>
    ))}
  </Wrapper>
);

LevelsList.defaultProps = {
  selectedLevelId: null,
  currentLevelId: null,
};

LevelsList.propTypes = {
  levels: PropTypes.array.isRequired,
  selectedLevelId: PropTypes.number,
  currentLevelId: PropTypes.number,
};

const mapStateToProps = state => ({
  selectedLevelId: state.levelManager.selectedLevelId,
  currentLevelId: state.levelManager.currentLevelId,
  levels: state.levelManager.levels,
});

export default connect(mapStateToProps)(LevelsList);
