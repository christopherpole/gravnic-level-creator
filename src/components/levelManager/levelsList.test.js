import React from 'react';
import { configure, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';

import { LevelsList } from './levelsList';
import testLevels from '../../data/testLevels';

configure({ adapter: new Adapter() });

describe('The level manager', () => {
  let props;

  beforeEach(() => {
    props = {
      levels: testLevels,
      selectedLevelId: testLevels[0].id,
      currentLevelId: testLevels[0].id,
      retrieveLevels: () => {},
    };
  });
  it('Renders without exploding', () => {
    const levelManager = shallow(<LevelsList {...props} />);

    expect(levelManager).toHaveLength(1);
  });

  it('Matches the current snapshot with no levels', () => {
    const levelManager = shallow(<LevelsList {...props} levels={[]} />);

    expect(toJson(levelManager)).toMatchSnapshot();
  });

  it('Matches the current snapshot with levels', () => {
    const levelManager = shallow(<LevelsList {...props} />);

    expect(toJson(levelManager)).toMatchSnapshot();
  });
});
