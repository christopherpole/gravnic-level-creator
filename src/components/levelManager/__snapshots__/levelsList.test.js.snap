// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The level manager Matches the current snapshot with levels 1`] = `
<styled.ul>
  <Connect(Component)
    id="1"
    isCurrent={true}
    isSelected={true}
    key="1"
    name="Test level 1"
    tiles={
      Array [
        Object {
          "position": 0,
          "selectedTileId": 2,
        },
        Object {
          "position": 1,
          "selectedTileId": 3,
        },
        Object {
          "position": 2,
          "selectedTileId": 0,
        },
      ]
    }
  >
    Test level 1
  </Connect(Component)>
  <Connect(Component)
    id="2"
    isCurrent={false}
    isSelected={false}
    key="2"
    name="Test level 2"
    tiles={
      Array [
        Object {
          "position": 0,
          "selectedTileId": 2,
        },
        Object {
          "position": 1,
          "selectedTileId": 3,
        },
        Object {
          "position": 2,
          "selectedTileId": 0,
        },
      ]
    }
  >
    Test level 2
  </Connect(Component)>
  <Connect(Component)
    id="3"
    isCurrent={false}
    isSelected={false}
    key="3"
    name="Test level 3"
    tiles={
      Array [
        Object {
          "position": 0,
          "selectedTileId": 2,
        },
        Object {
          "position": 1,
          "selectedTileId": 3,
        },
        Object {
          "position": 2,
          "selectedTileId": 0,
        },
      ]
    }
  >
    Test level 3
  </Connect(Component)>
</styled.ul>
`;

exports[`The level manager Matches the current snapshot with no levels 1`] = `<styled.ul />`;
